### PR TITLE
feat(api): added route to search across the models

### LINKED ISSUE #204: [3JH] Search engine View
## Describe the issue
This view should be used to match a keyword against multiple models from our database, namely:
- team
- mission
- vulnerability
- Users (optional)

The route should take `?q=<keyword>` as parameter and be able to search in an efficient way in other records. Optionally, it could take as parameter `&look_for_team=<True|False>` in order to limit calls to the database when those are not necessary

## Resource request
N/A

### PR DESCRIPTION
**Pull request by hamid-yg**
# Description

Search across the different models (q & look_for_team) 

# Fixes

#204 

- fixes #0

# Changes include

- [ ] Bugfix (non-breaking change that solves an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)

# Breaking changes

Please complete this section if any breaking changes have been made, otherwise delete it.

# Checklist

- [x] I have assigned this PR to myself
- [x] I have added @vrn-sh/members to the reviewers, or specific members of the team
- [x] I have added the needed labels
- [ ] I have tested this code
- [ ] I have added / updated tests (unit / functionals / end-to-end / ...)
- [x] I have updated the README and other relevant documents (guides...)

# Additional comments

It's not definitive. I will update this pr

### CODE REVIEW

File: api/api/views/viewsets/search.py
Comment (djnnvx) [inline api/api/views/viewsets/search.py]: You need to filter items based on if users can access them or not
(eg: a random pentester that is not member of team A should not access Team A's information)

File: api/api/views/viewsets/search.py
Comment (djnnvx) [inline api/api/views/viewsets/search.py]: it would be more convenient for the front-end if you could sort items by the closest match. This can be done trivially like this: 

```python
def get_name_value(self, obj) -> str:
        """
            get value used to compute most recurrent letters

            to do that, we get the important metadata that the user _should_ be
            looking for, & then how many letters are most similar. higher score wins.

        """
        if isinstance(obj, Manager):
            return f'{obj.first_name}{obj.last_name}{obj.email}'
        # you get the idea, ...


 def sort_by_closest_match(self, user_input: str, items: List[Any]) -> List[Any]:
        scores: dict[Any, int] = {}

        for item in items:
            score = 0

            # fetching the relevant metadata that the user would be
            # looking for (spaces or whatever dont matter)
            item_str = self.get_name_value(item) # You will have to implement this

            # all characters should be lowercase for a more accurate score
            if len(user_input) <= len(item_str):
                first_string, second_string = user_input.lower(), item_str.lower()
            else:
                first_string, second_string = item_str.lower(), user_input.lower()

            len_first = len(first_string)
            for ctr in range(len_first):
                score += 1 if first_string[ctr] == second_string[ctr] else 0

            scores[item] = score

        # sort scores by dict value (in reverse), then put in in a new dict
        sorted_scores = dict(sorted(scores.items(), key=lambda item: item[1], reverse=True))

        # just make a list out of the keys
        return list(sorted_scores.keys())
```

It can be even faster with numpy, but i don't have time to look into it right now

File: api/api/views/viewsets/search.py
Comment (djnnvx) [inline api/api/views/viewsets/search.py]: Instead of having a list of `filter()` statements that repeat themselves, you can do something like this: 

```python
def fetch(method, searchable_fields, keywords):
    """will perform search query in a generic way"""

    if not keywords:
        return None

    words = keywords.split()
    query = Q()

    for fields in searchable_fields:
        for word in words:

            # no SQLi for you :%)
            if not word.isalnum():
                    continue

            # some input validation is necessary before executing this
            query.add(Q(**{f'{fields}__icontains': word}), Q.OR)
    
    # is_allowed_to_see() is not implemented in this example, 
    # but im sure you get the idea
    return [x for x in method().filter(query) if is_allowed_to_see(user, x)]
```

Then, you can call it like this: 
```python
user_keywords = ['first_name', 'last_name', 'email']
search = fetch(request.user, Manager.objects.all, user_keywords, request.GET.get('q'))
```

Comment (djnnvx) [reply api/api/views/viewsets/search.py]: then, it's pretty easy to add a `search_managers=True|False` parameter to check if the request should be ran or not, for even better optimizations B)

**Review by djnnvx — CHANGES_REQUESTED** [review]
This looks like a good start (thank you for this awesome contribution), but it could be much more easy to maintain & to test; also there is no unit tests ): 

You should also add `drf_yasg`'s decorators to fix the documentation for front-end team

**Comment by djnnvx** [comment]
just thought about this, but query results also need to be paginated ? Database could easily be overload by someone searching with a single letter. Please add a `size` parameter that defaults to a fixed value, like 20 or 30.

**Comment by hamid-yg** [comment]
Still need to check if it works before merging

**Comment by djnnvx** [comment]
this looks much much better ! Well done :sparkles: 

do you think you can add unit tests for this feature ?

**Comment by djnnvx** [comment]
is this OK to merge ? (once `urls.py` conflict is fixed)

**Comment by hamid-yg** [comment]
@bogdzn it's working now. Check and merge it when you can please

**Pull request merged**
